File: api/app/utils/permission_defaults.py
"""
Default permission configurations for roles.
"""
from app.models.permission import Resource, Action, SystemRole
from app.models.organization import MemberRole


# 资源中文名称
RESOURCE_NAMES: dict[Resource, str] = {
    Resource.APARTMENT: "公寓管理",
    Resource.ROOM: "房间管理",
    Resource.TENANT: "租客管理",
    Resource.LEASE: "租约管理",
    Resource.BILL: "账单管理",
    Resource.UTILITY: "水电管理",
    Resource.MEMBER: "成员管理",
    Resource.SETTINGS: "系统设置",
    Resource.REPORT: "报表分析",
}

# 操作中文名称
ACTION_NAMES: dict[Action, str] = {
    Action.VIEW: "查看",
    Action.CREATE: "创建",
    Action.EDIT: "编辑",
    Action.DELETE: "删除",
    Action.EXPORT: "导出",
    Action.MANAGE: "管理（全部）",
}


def get_permission_name(resource: Resource, action: Action) -> str:
    """生成权限显示名称"""
    return f"{RESOURCE_NAMES.get(resource, resource.value)}{ACTION_NAMES.get(action, action.value)}"


# 组织角色默认权限配置
# owner 角色拥有所有权限，不可修改
DEFAULT_ORG_PERMISSIONS: dict[MemberRole, list[tuple[Resource, Action]]] = {
    MemberRole.OWNER: [
        # owner 拥有所有权限
        (Resource.APARTMENT, Action.VIEW),
        (Resource.APARTMENT, Action.CREATE),
        (Resource.APARTMENT, Action.EDIT),
        (Resource.APARTMENT, Action.DELETE),
        (Resource.ROOM, Action.VIEW),
        (Resource.ROOM, Action.CREATE),
        (Resource.ROOM, Action.EDIT),
        (Resource.ROOM, Action.DELETE),
        (Resource.TENANT, Action.VIEW),
        (Resource.TENANT, Action.CREATE),
        (Resource.TENANT, Action.EDIT),
        (Resource.TENANT, Action.DELETE),
        (Resource.LEASE, Action.VIEW),
        (Resource.LEASE, Action.CREATE),
        (Resource.LEASE, Action.EDIT),
        (Resource.LEASE, Action.DELETE),
        (Resource.BILL, Action.VIEW),
        (Resource.BILL, Action.CREATE),
        (Resource.BILL, Action.EDIT),
        (Resource.BILL, Action.DELETE),
        (Resource.BILL, Action.EXPORT),
        (Resource.UTILITY, Action.VIEW),
        (Resource.UTILITY, Action.CREATE),
        (Resource.UTILITY, Action.EDIT),
        (Resource.UTILITY, Action.DELETE),
        (Resource.MEMBER, Action.VIEW),
        (Resource.MEMBER, Action.CREATE),
        (Resource.MEMBER, Action.EDIT),
        (Resource.MEMBER, Action.DELETE),
        (Resource.SETTINGS, Action.VIEW),
        (Resource.SETTINGS, Action.EDIT),
        (Resource.REPORT, Action.VIEW),
        (Resource.REPORT, Action.EXPORT),
    ],
    MemberRole.ADMIN: [
        # admin 拥有除了删除组织和修改权限外的所有权限
        (Resource.APARTMENT, Action.VIEW),
        (Resource.APARTMENT, Action.CREATE),
        (Resource.APARTMENT, Action.EDIT),
        (Resource.APARTMENT, Action.DELETE),
        (Resource.ROOM, Action.VIEW),
        (Resource.ROOM, Action.CREATE),
        (Resource.ROOM, Action.EDIT),
        (Resource.ROOM, Action.DELETE),
        (Resource.TENANT, Action.VIEW),
        (Resource.TENANT, Action.CREATE),
        (Resource.TENANT, Action.EDIT),
        (Resource.TENANT, Action.DELETE),
        (Resource.LEASE, Action.VIEW),
        (Resource.LEASE, Action.CREATE),
        (Resource.LEASE, Action.EDIT),
        (Resource.LEASE, Action.DELETE),
        (Resource.BILL, Action.VIEW),
        (Resource.BILL, Action.CREATE),
        (Resource.BILL, Action.EDIT),
        (Resource.BILL, Action.DELETE),
        (Resource.BILL, Action.EXPORT),
        (Resource.UTILITY, Action.VIEW),
        (Resource.UTILITY, Action.CREATE),
        (Resource.UTILITY, Action.EDIT),
        (Resource.UTILITY, Action.DELETE),
        (Resource.MEMBER, Action.VIEW),
        (Resource.MEMBER, Action.CREATE),
        (Resource.MEMBER, Action.EDIT),
        # admin 不能删除成员
        (Resource.SETTINGS, Action.VIEW),
        (Resource.SETTINGS, Action.EDIT),
        (Resource.REPORT, Action.VIEW),
        (Resource.REPORT, Action.EXPORT),
    ],
    MemberRole.MEMBER: [
        # member 日常操作权限
        (Resource.APARTMENT, Action.VIEW),
        (Resource.ROOM, Action.VIEW),
        (Resource.ROOM, Action.EDIT),  # 可以修改房间状态
        (Resource.TENANT, Action.VIEW),
        (Resource.TENANT, Action.CREATE),
        (Resource.TENANT, Action.EDIT),
        (Resource.LEASE, Action.VIEW),
        (Resource.LEASE, Action.CREATE),
        (Resource.LEASE, Action.EDIT),
        (Resource.BILL, Action.VIEW),
        (Resource.BILL, Action.CREATE),
        (Resource.BILL, Action.EDIT),
        (Resource.UTILITY, Action.VIEW),
        (Resource.UTILITY, Action.CREATE),
        (Resource.UTILITY, Action.EDIT),
        (Resource.MEMBER, Action.VIEW),
        (Resource.SETTINGS, Action.VIEW),
        (Resource.REPORT, Action.VIEW),
    ],
    MemberRole.VIEWER: [
        # viewer 只读权限
        (Resource.APARTMENT, Action.VIEW),
        (Resource.ROOM, Action.VIEW),
        (Resource.TENANT, Action.VIEW),
        (Resource.LEASE, Action.VIEW),
        (Resource.BILL, Action.VIEW),
        (Resource.UTILITY, Action.VIEW),
        (Resource.MEMBER, Action.VIEW),
        (Resource.REPORT, Action.VIEW),
    ],
}


# 系统角色默认权限配置
DEFAULT_SYSTEM_ROLE_PERMISSIONS: dict[SystemRole, list[tuple[Resource, Action]]] = {
    SystemRole.SUPER_ADMIN: [
        # 超级管理员拥有所有权限
        (r, a)
        for r in Resource
        for a in [Action.VIEW, Action.CREATE, Action.EDIT, Action.DELETE, Action.EXPORT]
    ],
    SystemRole.SUPPORT: [
        # 客服：只读权限
        (Resource.APARTMENT, Action.VIEW),
        (Resource.ROOM, Action.VIEW),
        (Resource.TENANT, Action.VIEW),
        (Resource.TENANT, Action.EDIT),  # 可以修改租客信息
        (Resource.LEASE, Action.VIEW),
        (Resource.BILL, Action.VIEW),
        (Resource.UTILITY, Action.VIEW),
        (Resource.MEMBER, Action.VIEW),
        (Resource.REPORT, Action.VIEW),
    ],
    SystemRole.OPERATIONS: [
        # 运营：查看和导出权限
        (Resource.APARTMENT, Action.VIEW),
        (Resource.ROOM, Action.VIEW),
        (Resource.TENANT, Action.VIEW),
        (Resource.LEASE, Action.VIEW),
        (Resource.BILL, Action.VIEW),
        (Resource.BILL, Action.EXPORT),
        (Resource.UTILITY, Action.VIEW),
        (Resource.MEMBER, Action.VIEW),
        (Resource.SETTINGS, Action.VIEW),
        (Resource.REPORT, Action.VIEW),
        (Resource.REPORT, Action.EXPORT),
    ],
    SystemRole.FINANCE: [
        # 财务：账单和报表相关权限
        (Resource.BILL, Action.VIEW),
        (Resource.BILL, Action.CREATE),
        (Resource.BILL, Action.EDIT),
        (Resource.BILL, Action.EXPORT),
        (Resource.LEASE, Action.VIEW),
        (Resource.TENANT, Action.VIEW),
        (Resource.ROOM, Action.VIEW),
        (Resource.REPORT, Action.VIEW),
        (Resource.REPORT, Action.EXPORT),
    ],
    SystemRole.READONLY: [
        # 只读：只能查看
        (Resource.APARTMENT, Action.VIEW),
        (Resource.ROOM, Action.VIEW),
        (Resource.TENANT, Action.VIEW),
        (Resource.LEASE, Action.VIEW),
        (Resource.BILL, Action.VIEW),
        (Resource.UTILITY, Action.VIEW),
        (Resource.MEMBER, Action.VIEW),
    ],
}


# 系统角色配置
SYSTEM_ROLE_CONFIGS: list[dict] = [
    {
        "role": SystemRole.SUPER_ADMIN,
        "name": "超级管理员",
        "description": "拥有系统所有权限，可管理所有组织",
    },
    {
        "role": SystemRole.SUPPORT,
        "name": "客服",
        "description": "可查看所有组织数据，协助用户解决问题",
    },
    {
        "role": SystemRole.OPERATIONS,
        "name": "运营",
        "description": "可查看和导出数据，进行运营分析",
    },
    {
        "role": SystemRole.FINANCE,
        "name": "财务",
        "description": "可管理账单和查看财务报表",
    },
    {
        "role": SystemRole.READONLY,
        "name": "只读",
        "description": "只能查看数据，无修改权限",
    },
]

File: api/src/constants/permissionDefaults.ts
/**
 * 权限常量 — 静态定义，无需数据库存储
 */

import type { Resource, Action } from '@propease/api-contract';

// Re-export from api-contract
export { RESOURCES, ACTIONS, toPermissionCodes } from '@propease/api-contract';
export type { Resource, Action };

// 资源模块（无 settings）
export const RESOURCE_MODULES = ['apartment', 'room', 'tenant', 'lease', 'bill', 'utility', 'member', 'report'] as const;
export type ResourceModule = (typeof RESOURCE_MODULES)[number];

// 操作类型
export const ACTION_TYPES = ['view', 'create', 'edit', 'delete', 'export'] as const;
export type ActionType = (typeof ACTION_TYPES)[number];

// 预制角色名称
export const SYSTEM_ORG_ROLES = ['组织所有者', '公寓管理人', '一般合伙人'] as const;
export type SystemOrgRoleName = (typeof SYSTEM_ORG_ROLES)[number];

/**
 * 生成权限码
 */
export function toPermissionCode(resource: Resource, action: Action): string {
  return `${resource}:${action}`;
}

/**
 * 预制角色的默认权限码数组
 * 组织所有者：全部45个权限（9个模块 × 5个操作）
 * 公寓管理人：40个权限（8个模块 × 5个操作，无 settings）
 * 一般合伙人：16个权限（8个模块 × 2个操作 view/export，无 settings）
 */
export const DEFAULT_ORG_ROLE_PERMISSIONS: Record<SystemOrgRoleName, string[]> = {
  '组织所有者': (() => {
    const codes: string[] = [];
    for (const resource of ['apartment', 'room', 'tenant', 'lease', 'bill', 'utility', 'member', 'settings', 'report'] as Resource[]) {
      for (const action of ACTION_TYPES) {
        codes.push(toPermissionCode(resource, action));
      }
    }
    return codes;
  })(),
  '公寓管理人': (() => {
    const codes: string[] = [];
    for (const resource of RESOURCE_MODULES) {
      for (const action of ACTION_TYPES) {
        codes.push(toPermissionCode(resource, action));
      }
    }
    return codes;
  })(),
  '一般合伙人': (() => {
    const codes: string[] = [];
    for (const resource of RESOURCE_MODULES) {
      for (const action of ['view', 'export'] as Action[]) {
        codes.push(toPermissionCode(resource, action));
      }
    }
    return codes;
  })(),
};

/**
 * 预制角色定义
 */
export interface DefaultOrgRole {
  name: SystemOrgRoleName;
  description: string;
  is_system: true;
  permissions: string[];
}

export const DEFAULT_ORG_ROLES: DefaultOrgRole[] = [
  {
    name: '组织所有者',
    description: '拥有组织的全部权限，可管理所有功能和成员',
    is_system: true,
    permissions: DEFAULT_ORG_ROLE_PERMISSIONS['组织所有者'],
  },
  {
    name: '公寓管理人',
    description: '拥有除系统设置外的所有权限，可管理日常运营',
    is_system: true,
    permissions: DEFAULT_ORG_ROLE_PERMISSIONS['公寓管理人'],
  },
  {
    name: '一般合伙人',
    description: '拥有查看和导出权限，可查看数据但无法修改',
    is_system: true,
    permissions: DEFAULT_ORG_ROLE_PERMISSIONS['一般合伙人'],
  },
];
